host/traceh.c

```diff
@@ -70,10 +70,31 @@ void initialize_log_config()
         _log_level = _env2log_level();
         _log_file_name = getenv("OE_LOG_DEVICE");
         _custom_log_format = getenv("OE_LOG_FORMAT");
+        // check that custom log format string terminates with a line return
+        size_t len = strlen(_custom_log_format);
+        if (_custom_log_format[len - 1] != '\n')
+        {
+            fprintf(
+                stderr,
+                "%s\n",
+                "Custom log format does not end with a newline");
+            exit(1);
+        }
         _initialized = true;
     }
 }
 
+static void _replace_newline(char* log_msg)
```

Comment:
As discussed, this isn't enough to make it possible to write logs out as JSON event. Escaping messages is necessary.

Reply:
I would like to raise a discussion on what I am doing here. 

A lot of log messages seem to contain newlines. They possibly could contain single or double quotes or other characters that would need to be escaped. Apart from the fact that a custom log format might want to have one line per log message, which will break if we allow newline characters, if we want to support e.g. JSON for machine-readable purposes then other characters might cause further issues.

It could be that the user could indicate that they want to escape any characters on a log message and we could use some string library to do this, or do so manually.

For now I am just replacing newline characters with spaces, but please let me know if you would like me to invest some time in proposing a solution, or if you have any string handling mechanism in the codebase already.

Reply:
Replaced with a method called `_escape_characters` which does some basic control character escaping. I think this is sufficient for now but we can look into something more complete if we think it is necessary. I have made it so that it is called when the env variable `OE_JSON_ESCAPE` is set so that the user can choose if they want this functionality or not, depending on their log format